src/custom/state/orders/updater.ts

```diff
@@ -52,101 +52,180 @@ function _computeFulfilledSummary({
+      >(
+        (acc, { status, popupData }) => {
+          popupData && acc[status].push(popupData)
+          return acc
+        },
+        { fulfilled: [], expired: [], cancelled: [], unknown: [], pending: [] }
+      )
 
-      fulfillOrdersBatch({
-        ordersData,
-        chainId
-      })
+      // Bach state update per group, if any
+
+      fulfilled.length > 0 &&
+        fulfillOrdersBatch({
+          ordersData: fulfilled as OrderFulfillmentData[],
+          chainId
+        })
+      expired.length > 0 &&
+        expireOrdersBatch({
+          ids: expired as OrderID[],
+          chainId
+        })
+      cancelled.length > 0 &&
+        cancelOrdersBatch({
+          ids: cancelled as OrderID[],
+          chainId
+        })
     },
-    [fulfillOrdersBatch]
+    [cancelOrdersBatch, expireOrdersBatch, fulfillOrdersBatch]
   )
 
   useEffect(() => {
-    if (!chainId || pending.length <= 0) {
+    if (!chainId) {
       return
     }
 
-    const interval = setInterval(() => updateOrders(pending, chainId), OPERATOR_API_POLL_INTERVAL)
+    const interval = setInterval(() => updateOrders(chainId), OPERATOR_API_POLL_INTERVAL)
 
     return () => clearInterval(interval)
-  }, [pending, chainId, updateOrders])
+  }, [chainId, updateOrders])
 
   return null
 }
 
-export function ExpiredOrdersWatcher(): null {
+/**
+ * Updater for cancelled orders.
+ *
+ * Similar to Event updater, but instead of watching pending orders, it watches orders that have been cancelled
+ * in the last 5 min.
+ *
+ * Whenever an order that was cancelled but has since been fulfilled, trigger a state update
+ * and a popup notification, changing the status from cancelled to fulfilled.
+ *
+ * It's supposed to fix race conditions between the api accepting a cancellation while a solution was already
+ * submitted to the network by a solver.
+ * Due to the network's nature, we can't tell whether an order has been really cancelled, so we prefer to wait a short
+ * period and say it's cancelled even though in some cases it might actually be filled.
+ */
+export function CancelledOrdersUpdater(): null {
   const { chainId } = useActiveWeb3React()
 
-  const expireOrdersBatch = useExpireOrdersBatch()
+  const cancelled = useCancelledOrders({ chainId })
 
-  const pendingOrders = usePendingOrders({ chainId })
+  // Ref, so we don't rerun useEffect
+  const cancelledRef = useRef(cancelled)
+  cancelledRef.current = cancelled
 
-  // ref, so we don't rerun useEffect
-  const pendingOrdersRef = useRef(pendingOrders)
-  pendingOrdersRef.current = pendingOrders
+  const fulfillOrdersBatch = useFulfillOrdersBatch()
 
-  useEffect(() => {
-    if (!chainId) return
+  const updateOrders = useCallback(
+    async (chainId: ChainId) => {
+      // Filter orders created in the last 5 min, no further
+      const pending = cancelledRef.current.filter(
+        order => Date.now() - new Date(order.creationTime).getTime() < CANCELLED_ORDERS_PENDING_TIME
```

Comment:
I didn't realised this, maybe is not that important. But the min you check it with the creation time, not with the deletion time. 

Actually, this might solve the issue you mentioned that you have here. You compare times of two different clocks, the server and the client. If you save the deletion time, then it's easier for you to count the 5min reliably. 

Anyways, even with this approach is not optimal. The ideal would be if the backend could tell us if they managed to delete the order (meaning, no settlement is in process of sending the order).

No need to change anything I think, Im just thinking aloud.

Reply:
Yes, this is what I mentioned the other day.
I haven't thought about storing it locally though (:facepalm:)

I could do that indeed, but as you mentioned, wouldn't change the fact that we can't really tell when an order is in progress of being executed.